---
title: "Benchmarking tidyomics efficiency"
author:
- name: William Hutchison
  affiliation: Walter and Eliza Hall Institute of Medical Research
  email: hutchison.w@wehi.edu.au
- name: Timothy Keyes
  affiliation: Stanford University
  email: tkeyes@stanford.edu
package: 
output:
  BiocStyle::html_document
abstract: |
  Evaluate the efficiency of common data analysis operations with both standard and tidyomics workflows.
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Set up
```{r, message = FALSE, warning = FALSE, results = FALSE}
# Load libraries
library(Seurat)
library(SummarizedExperiment)
library(edgeR)
library(tidyseurat)
library(tidySingleCellExperiment)
library(tidySpatialExperiment)
library(tidySummarizedExperiment)
library(tidybulk)
library(plyranges)
library(tidytof)
library(purrr)
library(ggplot2)
library(tibble)
library(dplyr)
library(microbenchmark)
library(cowplot)

# Load data
seurat_object <- tidyomicsWorkshop::seurat_obj
sce_object <- tidyomicsWorkshop::sce_obj
example(read10xVisium)
spe_object <- spe
se_object <- counts_SE
samusik_flowset <- HDCytoData::Samusik_01_flowSet()
samusik_flowframe <- samusik_flowset[[1]]
samusik_tibble <- 
  tidytof::as_tof_tbl(samusik_flowset) |> 
  mutate(population_id = as.character(population_id))
```

# Create benchmarking functions
## tidyseurat 

Aggregate cells
```{r}
# tidyomics workflow
seurat_tidy_aggregate <- 
  function() {
    seurat_object |>
      tidyseurat::aggregate_cells(batch)
  }

# Standard workflow
seurat_base_aggregate <- 
  function() {
    seurat_object_split <- SplitObject(seurat_object, split.by = "batch")
    seurat_object_split_counts <- lapply(seurat_object_split, GetAssayData, slot = "counts")
    lapply(seurat_object_split_counts, rowSums)
  }
```

Plot feature counts per cell
```{r}
# tidyomics workflow
seurat_tidy_plot <-
  function() {
    seurat_object |>
      ggplot(aes(x = batch, y = nCount_RNA)) +
      geom_boxplot()
  }

# Standard workflow
seurat_base_plot <-
  function() {
    seurat_object_split <- SplitObject(seurat_object, split.by = "batch")
    seurat_object_split_counts <- lapply(seurat_object_split, GetAssayData, slot = "counts")
    sce_object_split_counts_sums <- lapply(seurat_object_split_counts, colSums)
    boxplot(sce_object_split_counts_sums)
  }
```

Calculate and filter markers based on an aggregate gene signature
```{r}
# tidyomics workflow
seurat_tidy_signature <-
  function() {
    seurat_object |>
      join_features(features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide") |>
      tidyseurat::mutate(signature_score =
               scales::rescale(CD3D + TRDC + TRGC1+ TRGC2, to=c(0,1)) -
               scales::rescale(CD8A + CD8B, to=c(0,1))
      ) |>
      tidyseurat::filter(signature_score > 0.7)
  }

# Standard workflow
seurat_base_signature <-
  function() {
    counts_positive <-
      seurat_object[c("CD3D", "TRDC", "TRGC1", "TRGC2"), ] |>
      colSums() |>
      scales::rescale(to = c(0, 1))
    counts_negative <-
      seurat_object[c("CD8A", "CD8B"), ] |>
      colSums() |>
      scales::rescale(to = c(0, 1))
    seurat_object$signature_score <- counts_positive - counts_negative
    seurat_object[, seurat_object$signature_score > 0.7]
  }
```

## tidySingleCellExperiment
Aggregate cells (within each batch)
```{r}
# tidyomics workflow
sce_tidy_aggregate <-
  function() {
    sce_object |>
      aggregate_cells(batch)
  }

# Standard workflow
sce_base_aggregate <- 
  function() {
    sce_object_split <- split(sce_object, sce_object$batch)
    sce_object_split_counts <- lapply(sce_object_split, counts)
    lapply(sce_object_split_counts, rowSums)
  }
```

Plot feature counts per cell
```{r}
# tidyomics workflow
sce_tidy_plot <-
  function() {
    sce_object |>
      ggplot(aes(x = batch, y = nCount_RNA)) +
      geom_boxplot()
  }

# Standard workflow
sce_base_plot <-
  function() {
    sce_object_split <- split(sce_object, sce_object$batch)
    sce_object_split_counts <- lapply(sce_object_split, counts)
    sce_object_split_counts_sums <- lapply(sce_object_split_counts, colSums)
    boxplot(sce_object_split_counts_sums)
  }
```

Subset cells by feature signature
```{r}
# tidyomics workflow
sce_tidy_signature <- 
  function() {
    sce_object |>
      join_features(features = c("CD3D", "TRDC", "TRGC1", "TRGC2", "CD8A", "CD8B"), shape = "wide") |>
      mutate(signature_score =
               scales::rescale(CD3D + TRDC + TRGC1 + TRGC2, to = c(0, 1)) -
               scales::rescale(CD8A + CD8B, to = c(0, 1))
      ) |>
      tidySingleCellExperiment::filter(signature_score > 0.7)
  }

# Standard workflow
sce_base_signature <- 
  function() {
    counts_positive <-
      assay(sce_object, "logcounts")[c("CD3D", "TRDC", "TRGC1", "TRGC2"), ] |>
      colSums() |>
      scales::rescale(to = c(0, 1))
    counts_negative <-
      assay(sce_object, "logcounts")[c("CD8A", "CD8B"), ] |>
      colSums() |>
      scales::rescale(to = c(0, 1))
    sce_object$signature_score <- counts_positive - counts_negative
    sce_object[, sce_object$signature_score > 0.7]
  }
```

## tidySpatialExperiment
Aggregate cells
```{r}
# tidyomics workflow
spe_tidy_aggregate <- 
  function() {
    spe_object |> 
      aggregate_cells(sample_id)
  }

# Standard workflow
spe_base_aggregate <- 
  function() {
    spe_object_split <- split(spe_object, spe_object$sample_id)
    spe_object_split_counts <- lapply(spe_object_split, counts)
    lapply(spe_object_split_counts, rowSums)
  }
```

Plot feature counts per cell
```{r}
# tidyomics workflow
spe_tidy_plot <- 
  function() {
    spe_object |>
      join_features(spe_object |> counts() |> rownames(), shape = "long") |>
      group_by(.cell, sample_id) |> 
      summarise(cell_sums = sum(.abundance_counts)) |>
      ggplot(aes(x = sample_id, y = cell_sums, group = sample_id)) +
      geom_boxplot()
  }

# Standard workflow
spe_base_plot <- 
  function() {
    spe_object_split <- split(spe_object, spe_object$sample_id)
    spe_object_split_counts <- lapply(spe_object_split, counts)
    spe_object_split_counts_sums <- lapply(spe_object_split_counts, colSums)
    boxplot(spe_object_split_counts_sums)
  }
```

## tidySummarizedExperiment 
Subset features by mean count
```{r}
# tidyomics workflow
se_tidy_mean <- function() {
  se_object |>
    group_by(.feature) |>
    mutate(mean_count = mean(count)) |>
    tidySummarizedExperiment::filter(mean_count > 0)
}

# Standard workflow
se_base_mean <- function() {
  se_object_count <- assay(se_object, "count")
  se_object_count_means <- rowMeans(se_object_count)
  se_object_count[se_object_count_means > 500, ]
}
```

Plot features per sample
```{r}
# tidyomics workflow
se_tidy_plot <- function() {
  se_object |>
    ggplot(aes(x = condition, y = count)) +
    geom_boxplot()
}

# Standard workflow
se_base_plot <- function() {
  se_object_count <- assay(se_object, "count")
  boxplot(se_object_count)
}
```

## tidybulk
Normalise
```{r message = FALSE}
# tidyomics workflow
bulk_tidy_normalise <- function() {
  se_object |>
    identify_abundant(factor_of_interest = condition) |>
    scale_abundance()
}

# Standard workflow
bulk_base_normalise <- function() {
  se_object_counts <- assay(se_object, "count")
  se_object_col_data <- colData(se_object)
  dge <- DGEList(counts = se_object_counts, 
                 sample = se_object_col_data, 
                 group = se_object_col_data$condition)
  dge <- calcNormFactors(dge, method = "TMM")
  edgeR::cpm(dge, log = TRUE, prior.count = 0.5)
}
```

Prepare data for next benchmark without timing
```{r message = FALSE}
se_object_counts <- assay(se_object, "count")
se_object_col_data <- colData(se_object)
dge <- DGEList(counts = se_object_counts, 
               sample = se_object_col_data, 
               group = se_object_col_data$condition)
dge <- calcNormFactors(dge, method = "TMM")

se_object_tidy_normalised <-
  se_object |>
  identify_abundant(factor_of_interest = condition) |>
  scale_abundance()

se_object_base_normalise <- 
  edgeR::cpm(dge, log = TRUE, prior.count = 0.5)
```

Plot effects of normalisation
```{r}
# tidyomics workflow
bulk_tidy_plot <- function() {
  se_object_tidy_normalised |>
    ggplot(aes(count_scaled + 1, group=sample, color=`Cell.type`)) +
    geom_density() +
    scale_x_log10()
}

# Standard workflow
bulk_base_plot <- function() {
  plot(density(se_object_base_normalise[, 1]), type = "n")
  n <-  dim(se_object_base_normalise)[2] - 1
  for(i in 1:n){
    lines(density(c(se_object_base_normalise[, i])))
  }
}
```

Identify most variable features per cell type
```{r}
# tidyomics workflow
bulk_tidy_variable <- 
  function() {
    se_object |>
      nest(data = -Cell.type) |>
      mutate(top_features = map(data, ~ 
        .x |>
          keep_variable(top = 10) |>
          rownames()
      ))
  }

# Standard workflow
bulk_base_variable <- 
  function() {
    top_features <- list()
    se_object_split <- split(se_object, se_object$Cell.type)
    for (i in 1:length(se_object_split)) {
      se_object_counts <- assay(se_object_split[[i]], "count")
      feature_variance <- apply(se_object_counts, 1, var)  
      top_features <- c(top_features, list(feature_variance[order(feature_variance)[1:5]]))
    }
  }
```

## plyranges
Prepare data without timing
```{r}
gwas <-
  data.frame(
    seqnames = 1,
    start = round(runif(100, 0, 100)),
    width = 1,
    rsID = paste0("rs", 1:100)
  ) |>
  as_granges()

exons <- data.frame(
  seqnames = 1,
  start = round(runif(100, 0, 100)),
  width = 5,
  exonID = paste0("e", 1:100)
) |>
  as_granges()

x <-
  data.frame(
    seqnames = 1,
    start = round(runif(100, 0, 1e4)),
    width = round(runif(100, 5, 15))
  ) |>
  as_granges() |>
  sort() |>
  mutate(xID = paste0("x", 1:100),
         group = paste0("g", rep(1:2, each = 50)))

y <-
  data.frame(
    seqnames = 1,
    start = round(runif(100, 0, 1e4)),
    width = round(runif(100, 5, 15))
  ) |>
  as_granges() |>
  sort() |>
  mutate(yID = paste0("y", 1:100))
```

Find and aggregate overlapping regions
```{r}
# tidyomics workflow
genome_tidy_aggregate <- function() {
  exons |>
    plyranges::join_overlap_inner(gwas) |>
    plyranges::group_by(rsID) |>
    plyranges::summarise(n = plyranges::n_distinct(exonID))
}

# Standard workflow
genome_base_aggregate <- function() {
  hits <- findOverlaps(exons, gwas, ignore.strand = FALSE)
  olap <-splitAsList(exons$exonID[queryHits(hits)], gwas$rsID[subjectHits(hits)])
  n <- lengths(unique(olap))
  DataFrame(rsID = names(n), n = as.integer(n))
}
```

Calculate and plot distance between sets of features
```{r}
# tidyomics workflow
genome_tidy_distance <- function() {
  x |>
    anchor_5p() |>
    mutate(width = 1) |>
    add_nearest_distance(y = y |> anchor_center() |> mutate(width = 1)) |>
    as_tibble() |>
    ggplot(aes(distance, group = group, fill = group)) +
    geom_histogram(position = "dodge")
}

# Standard workflow
genome_base_distance <- function() {
  x_5p <- resize(x, width = 1)
  y_mid <- y - ifelse(width(y) %% 2 == 0, width(y) / 2 - .5, floor(width(y) / 2))
  hits <- distanceToNearest(x_5p, y_mid)
  x$distance[queryHits(hits)] <- mcols(hits)$distance
  mcols(x)[, c("group", "distance")] |>
    as_tibble() |>
    ggplot(aes(distance, group = group, fill = group)) +
    geom_histogram(position = "dodge")
}
```

Subset areas of overlap between feature sets
```{r}
# tidyomics workflow
genome_tidy_overlap <- function() {
  x |>
    join_overlap_inner(range(x) |>
                         tile_ranges(width = 1000) %>%
                         mutate(tile = seq_along(.))) |>
    plyranges::group_by(tile) |>
    plyranges::disjoin_ranges(total = plyranges::n()) |>
    plyranges::filter(total > 1)
}

# Standard workflow
genome_base_overlap <- function() {
  tiles <- tile(range(x), width = 1000)[[1]]
  tiles$tile <- seq_along(tiles)
  hits <- findOverlaps(x, tiles)
  res <- lapply(1:length(tiles), function(t) {
    x_sub <- x[queryHits(hits)[subjectHits(hits) == t]]
    d <- disjoin(x_sub)
    cov <- as(coverage(x_sub), "GRanges")
    d[d %over% cov[cov$score > 1]]
  })
  do.call(c, res)
}
```


## tidytof
Downsample cells by cluster (aka by group)
```{r}
# tidyomics workflow
cytometry_tidy_downsample <-
  function() {
    samusik_tibble |>
      tidytof::tof_downsample_constant(group_cols = population_id, num_cells = 200L)
  }

# Standard workflow
cytometry_base_downsample <-
  function() {
    subset_flowframe <- function(flowframe) {
      num_cells <- nrow(flowframe)
      if (num_cells > 200L) {sample_indices <- sample(x = 1:num_cells, size = 200L)}
      else {sample_indices <- sample(x = 1:num_cells, size = num_cells)}
      subsampled_flowframe <- flowframe[sample_indices]
    }
    samusik_split <- split(samusik_flowframe,
                           f = as.factor(samusik_flowframe[, "population_id"]@exprs),
                           flowSet = TRUE)
    subsampled_flowset <- flowCore::fsApply(x = samusik_split, FUN = subset_flowframe)
    subsampled_flowframe <- as(subsampled_flowset, "flowFrame")
  }

```

Aggregate features by variable 
```{r}
# tidyomics workflow
cytometry_tidy_aggregate <- 
  function() {
    samusik_tibble |>
      tof_extract_central_tendency(cluster_col = population_id, 
                                   marker_cols = starts_with("CD"), 
                                   format = "long")
  }

# Standard workflow
cytometry_base_aggregate <- 
  function() {
    cd_cols <- grepl(pattern = "^CD", x = samusik_flowframe@description$MARKER_INFO$marker_name)
    aggregate_flowframe <- function(flowframe) {
      medians <- colMedians(flowframe[, cd_cols]@exprs)
      marker_names <- flowframe@description$MARKER_INFO$marker_name[cd_cols]
      group_ids <- unique(flowframe@exprs[, "population_id"])
      aggregated_result <- data.frame(marker_name = marker_names, 
                                      median = colMedians(flowframe[, cd_cols]@exprs), 
                                      group_id = group_ids)
    }
    samusik_split <- split(samusik_flowframe, 
                           f = as.factor(samusik_flowframe[, "population_id"]@exprs),
                           flowSet = TRUE)
    aggregated_list <- flowCore::fsApply(x = samusik_split, FUN = aggregate_flowframe)
    do.call("rbind", aggregated_list)
  }
```

Normalize and rescale 
```{r}
# tidyomics workflow
cytometry_tidy_normalise <- function() { 
  samusik_tibble |> 
    tof_preprocess() |> 
    tof_transform(transform_fun = scales::rescale)
}

# Standard workflow
cytometry_base_normalise <- function() { 
  result <- samusik_flowframe
  asinh_transformation <- arcsinhTransform(a = 0, b = 1/5)
  transform_list <- transformList(from = colnames(result), tfun = asinh_transformation)
  result <- transform(result, transform_list)
  result@exprs <- apply(result@exprs, MARGIN = 2, FUN = scales::rescale)
  return(result)
}
```

Prepare data without timing
```{r, warning = FALSE, results = FALSE}
samusik_tibble_pca <- 
  samusik_tibble |> 
  tof_preprocess() |> 
  tof_transform(transform_fun = scales::rescale) |> 
  dplyr::filter(population_id != "25")

normalized_flowframe <- cytometry_base_normalise()

samusik_split <- 
  flowCore::split(
    normalized_flowframe, 
    f = as.factor(normalized_flowframe[, "population_id"]@exprs), 
    flowSet = TRUE
  )
samusik_flowframe_pca <- as(samusik_split[1:24], "flowFrame")

```

Subset cells by feature signature 
```{r}
# tidyomics workflow
cytometry_tidy_signature <- function() { 
  samusik_tibble |> 
    mutate(signature = scales::rescale(`CD8|Eu151` + `CD3|Nd144`)) |> 
    dplyr::filter(signature > 0.3)
}

# Standard workflow
cytometry_base_signature <- function() { 
  result <- samusik_flowframe
  cols_to_sum <- grepl(pattern = "CD8$|CD3$", 
                       x = samusik_flowframe@description$MARKER_INFO$marker_name)
  signature <- samusik_flowframe@exprs[, cols_to_sum] |> 
    rowSums() |> 
    scales::rescale()
  result <- fr_append_cols(result, cols = matrix(data = signature, 
                                                 nrow = nrow(result), 
                                                 dimnames = list(NULL, "signature")))
  cells_to_keep <- signature > 0.3
  result[cells_to_keep, ]
}
```

Calculate and plot pca 
```{r}
# tidyomics workflow
cytometry_tidy_plot <- function() { 
  samusik_tibble_pca |> 
    tof_reduce_dimensions(pca_cols = starts_with("CD"), method = "pca") |> 
    tof_plot_cells_embedding(embedding_cols = c(.pc1, .pc2), color_col = population_id)
}

# Standard workflow
cytometry_base_plot <- function() { 
  cd_cols <- grepl(pattern = "^CD", x = samusik_flowframe@description$MARKER_INFO$marker_name)
  pca_exprs <- samusik_flowframe_pca@exprs[, cd_cols]
  pca_object <- prcomp(x = (pca_exprs))
  plot(x = pca_object$x[, 1], 
       y = pca_object$x[, 2], 
       col = as.factor(samusik_flowframe_pca@exprs[, "population_id"]))
}
```

# Summarise results

## Construct vectors
```{r, results = FALSE, warning = FALSE, message = FALSE, fig.keep = "none"}
package <-
  c(
    rep("tidyseurat", 3), 
    rep("tidySingleCellExperiment", 3),
    rep("tidySpatialExperiment", 2),
    rep("tidySummarizedExperiment", 2),
    rep("tidybulk", 3),
    rep("plyranges", 3),
    rep("tidytof", 5)
  ) |>
  rep(each = 2)

operation <- 
  c(
    "Aggregate cells by variable", "Plot feature counts per cell", "Subset cells by feature",
    "Aggregate cells by variable", "Plot feature counts per cell", "Subset cells by feature",
    "Aggregate cells by variable", "Plot feature counts per cell",
    "Subset features by mean count", "Plot features per sample",
    "Normalise feature", "Plot normalised feature density", "Find most variable features", 
    "Aggregate overlaps", "Plot feature set distances", "Grouped disjoin ranges",
     "Downsample cells", "Aggregate cells by variable", "Normalise features", "Subset cells by feature", "Plot PCA" 
  ) |>
  rep(each = 2)


method <-
  c("tidyomics", "standard") |>
  rep(times = 42 / 2)

metric <-
  c("Line count", "Variable count", "Time elapsed") |>
  rep(times = 42)

variable_count <- 
  c(
    0, 2, 0, 3, 0, 2, # tidySeurat
    0, 2, 0, 3, 0, 2, # tidySingleCellExperiment
    0, 2, 0, 3, # tidySpatialExperiment
    0, 2, 0, 1, # tidySummarizedExperiment
    0, 3, 0, 1, 0, 4, # tidybulk
    0, 3, 0, 3, 0, 6, # plyranges
    0, 6, 0, 7, 0, 3, 0, 4, 0, 3 # tidytof
  )

line_count <- 
  c(
    2, 3, 3, 4, 7, 10, # tidySeurat
    2, 3, 3, 4, 7, 10, # tidySingleCellExperiment
    2, 3, 6, 4, # tidySpatialExperiment
    4, 3, 3, 2, # tidySummarizedExperiment
    3, 7, 4, 5, 7, 7, # tidybulk
    4, 4, 7, 8, 7, 10, # plyranges
    2, 11, 4, 14, 3, 6, 3, 11, 3, 6 # tidytof
  )

time_elapsed <-
  microbenchmark(
    seurat_tidy_aggregate(),
    seurat_base_aggregate(),
    seurat_tidy_plot(),
    seurat_base_plot(),
    seurat_tidy_signature(),
    seurat_base_signature(),
    sce_tidy_aggregate(),
    sce_base_aggregate(),
    sce_tidy_plot(),
    sce_base_plot(),
    sce_tidy_signature(),
    sce_base_signature(),
    spe_tidy_aggregate(),
    spe_base_aggregate(),
    spe_tidy_plot(),
    spe_base_plot(),
    se_tidy_mean(),
    se_base_mean(),
    se_tidy_plot(),
    se_base_plot(),
    bulk_tidy_normalise(),
    bulk_base_normalise(),
    bulk_tidy_plot(),
    bulk_base_plot(),
    bulk_tidy_variable(),
    bulk_base_variable(),
    genome_tidy_aggregate(),
    genome_base_aggregate(),
    genome_tidy_distance(),
    genome_base_distance(),
    genome_tidy_overlap(),
    genome_base_overlap(),
    cytometry_tidy_downsample(),
    cytometry_base_downsample(),
    cytometry_tidy_aggregate(),
    cytometry_base_aggregate(),
    cytometry_tidy_normalise(),
    cytometry_base_normalise(),
    cytometry_tidy_signature(),
    cytometry_base_signature(),
    cytometry_tidy_plot(),
    cytometry_base_plot(),
    times = 100
  ) |>
  base::summary() |>
  pull(mean)
```

## Combine vectors
```{r, results = FALSE, warning = FALSE, message = FALSE}
# Zip line count, variable count and time elapsed together into a single vector
value <- c()
for (i in 1:length(line_count)) {
  value <- c(value, line_count[i], variable_count[i], time_elapsed[i])
}

# Create tibble 
tidy_base_benchmark <- 
  tibble(
    package, 
    method,
    operation
  ) |>
  
  # Replicate each row 3 times in place to allow joining of metrics in long format
  slice(rep(1:dplyr::n(), each = 3)) |>
  add_column(metric, value) |>

  # Adjust tibble for plotting 
  mutate(
    package = factor(package, levels = c("plyranges", "tidySingleCellExperiment", "tidySpatialExperiment", "tidySummarizedExperiment", "tidybulk", "tidytof", "tidyseurat")), 
    metric = if_else(metric == "Time elapsed", "Time elapsed\n(ms)", paste0(metric, "\n")), 
    metric = factor(metric, levels = c("Variable count\n", "Line count\n", "Time elapsed\n(ms)"))
  )
```

## Plot
```{r, warning = FALSE}
# Create theme for plotting
theme_multipanel <-
  theme_bw() +
  theme(
    panel.border = element_blank(),
    axis.line = element_line(linewidth = 0.1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    strip.background = element_blank(),
    axis.title.y = element_text(margin = margin(
      t = 0,
      r = 0,
      b = 0,
      l = 0
    ), size = 7),
    axis.title.x = element_text(margin = margin(
      t = 0,
      r = 0,
      b = 0,
      l = 0
    ), size = 7),
    panel.spacing.x = unit(0.05, "lines"),
    panel.spacing.y = unit(0.25, "lines"),
    axis.text.x = element_text(size = 5),
    axis.text.y = element_text(size = 5),
    strip.text.x = element_text(size = 4.5, face = "bold"),
    strip.text.y = element_text(size = 5),
    strip.clip = "off",

    # legend
    legend.key.size = unit(5, 'mm'),
    legend.title = element_text(size = 5),
    legend.text = element_text(size = 5),
    
    # Title
    plot.title = element_text(size = 7),
    axis.line.x = element_line(size = 0.2),
    axis.line.y = element_line(size = 0.2),
    axis.ticks.x = element_line(size = 0.2),
    axis.ticks.y = element_line(size = 0.2)
  )
```

```{r, warning = FALSE}
# Plot each tidyomics and base metrics separately
tidy_base_benchmark |>
  ggplot(aes(
    x = operation,
    y = value,
    group = method,
    colour = method
  )) +
  geom_line() +
  geom_point(size = 1) +
  facet_grid(
    rows = vars(metric),
    cols = vars(package),
    scale = "free") +
  theme_multipanel +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1), add = c(1, 0))) + 
  scale_colour_manual(values = c("#56B4E9", "#B14380")) +
  theme(
    axis.text.x = element_text(
      angle = 45,
      vjust = 1,
      hjust = 1,
    ),
    plot.margin = margin(l = 3),
    legend.position = "right") +
  labs(
    x = "Operation",
    y = NULL,
    colour = "Method"
  )
```

## Calculate summary statistics 
```{r, message = FALSE}
tidy_base_benchmark_summaries <-
  tidy_base_benchmark |>
  group_by(method, metric) |>
  summarise(method_mean = mean(value))

standard_line_mean <- 
  tidy_base_benchmark_summaries |> 
  dplyr::filter(metric == "Line count\n" & method == "standard") |>
  pull(method_mean)

tidy_line_mean <- 
  tidy_base_benchmark_summaries |> 
  dplyr::filter(metric == "Line count\n" & method == "tidyomics") |>
  pull(method_mean)

standard_time_mean <- 
  tidy_base_benchmark_summaries |> 
  dplyr::filter(metric == "Time elapsed\n(ms)" & method == "standard") |>
  pull(method_mean)

tidy_time_mean <- 
  tidy_base_benchmark_summaries |> 
  dplyr::filter(metric == "Time elapsed\n(ms)" & method == "tidyomics") |>
  pull(method_mean)

# Line count percentage change
(standard_line_mean - tidy_line_mean) / standard_line_mean * 100

# Time elapsed percentage change
(standard_time_mean - tidy_time_mean) / standard_time_mean * 100
```

# Session info

```{r sessionInfo, echo = FALSE}
sessionInfo()
```
